pkg/pillar/cmd/zedagent/handleProfile.go

```diff
@@ -0,0 +1,326 @@
+	"github.com/lf-edge/eve/pkg/pillar/flextimer"
+	"github.com/lf-edge/eve/pkg/pillar/types"
+	"github.com/lf-edge/eve/pkg/pillar/zedcloud"
+)
+
+const (
+	defaultLocalProfileServerPort = "8888"
+	savedLocalProfileFile         = "lastlocalprofile"
+)
+
+// urlAndSrcIP structure for mapping source IP and url of local server
+type urlAndSrcIP struct {
+	srcIP     net.IP
+	actualURL string
+}
+
+func getLocalProfileURL(localProfileServer string) (string, error) {
+	localProfileURL := fmt.Sprintf("http://%s", localProfileServer)
+	u, err := url.Parse(localProfileURL)
+	if err != nil {
+		return "", fmt.Errorf("url.Parse: %s", err)
+	}
+	if u.Port() == "" {
+		localProfileURL = fmt.Sprintf("%s:%s", localProfileURL, defaultLocalProfileServerPort)
+	}
+	return fmt.Sprintf("%s/api/v1/local_profile", localProfileURL), nil
+}
+
+//checkAndHandleLocalProfileUpdate gets local profile config if localProfileServer defined,
+//determines currentProfile and notify on changes
+func checkAndHandleLocalProfileUpdate(getconfigCtx *getconfigContext) {
+	localProfileServer := getconfigCtx.localProfileServer
+	if localProfileServer == "" {
+		return
+	}
+	localProfileURL, err := getLocalProfileURL(localProfileServer)
+	if err != nil {
+		log.Errorf("getLocalProfileURL: %s", err)
+	} else {
+		if localProfileObject, err := getLocalProfileConfig(localProfileURL, getconfigCtx); err != nil {
+			log.Errorf("localProfileTimerTask getLocalProfileConfig: %s", err)
+		} else {
+			localProfile := localProfileObject.GetLocalProfile()
+			if getconfigCtx.localProfile != localProfile {
+				log.Noticef("localProfileTimerTask: local profile changed from %s to %s",
+					getconfigCtx.localProfile, localProfile)
+				getconfigCtx.localProfile = localProfile
+			}
+			currentProfile := determineCurrentProfile(getconfigCtx)
+			if getconfigCtx.currentProfile != currentProfile {
+				log.Noticef("localProfileTimerTask: profile changed from %s to %s",
+					getconfigCtx.currentProfile, currentProfile)
+				getconfigCtx.currentProfile = currentProfile
+				publishZedAgentStatus(getconfigCtx)
+			}
+			writeReceivedProtoMessageLocalProfile(localProfileObject)
+		}
+	}
+}
+
+// Run a periodic fetch of the currentProfile
+func localProfileTimerTask(handleChannel chan interface{}, getconfigCtx *getconfigContext) {
+
+	ctx := getconfigCtx.zedagentCtx
+
+	// use ConfigInterval as localProfileInterval
+	localProfileInterval := ctx.globalConfig.GlobalValueInt(types.ConfigInterval)
+	interval := time.Duration(localProfileInterval) * time.Second
+	max := float64(interval)
+	min := max * 0.3
+	ticker := flextimer.NewRangeTicker(time.Duration(min),
+		time.Duration(max))
+	// Return handle to caller
+	handleChannel <- ticker
+
+	wdName := agentName + "currentProfile"
+
+	// Run a periodic timer so we always update StillRunning
+	stillRunning := time.NewTicker(25 * time.Second)
+	ctx.ps.StillRunning(wdName, warningTime, errorTime)
+	ctx.ps.RegisterFileWatchdog(wdName)
+
+	for {
```

Comment:
We talked about doing this differently by starting without the periodic timer and only do a select (or channel read) for the trigger.
Once the trigger has hit once we can proceed since we parseProfile will have been called and we know 1) the globalProfile and 2) the localProfileServer.

If the localProfileServer is set we should read and parse the saved localProfile, otherwise we should clear the local profile file. Then start the timer and do this loop.

Reply:
Yes, but after observe the code I see, that we can have situation, when we call parseProfile before actually starting the timerProfile function (receiver of notify), so our trigger will do nothing in some cases.
If you think, that it is reasonable to clean saved profile here, not in parseProfile function, I can add ticker here checking every second if we called parseProfile function.

Reply:
Why can't you create the channel (len 1) in the task up front, so that trigger would always work?

Reply:
We cannot select from buffered channel in for loop in this case.

Reply:
If you create the channel (length 1) up front then the notify will always work, right?

Reply:
If we know that we will get an initial notification the first time parseProfile is called, then we can just read once from the channel.

But I think we actually need to know the localProfile from the saved file before then.
Why can't we read the file at the top of the task and set ctx.localProfile based on that?

Reply:
We can do it before starting of parsing any configs (before run of gouroutine timer for edgeDevConfig) and just set local profile (it can be cleared later with empty local profile server). What do you think about this variant?

Reply:
That is what I'm suggesting.